sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/types/Arguments.kt

```diff
@@ -0,0 +1,99 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.types
+
+import com.squareup.sqldelight.SqliteParser
+import com.squareup.sqldelight.resolution.query.Value
+import java.util.ArrayList
+import java.util.LinkedHashSet
+
+/**
+ * Represents the type for an argument. All that is contained is a value since the origin of the
+ * type may need to be known for adapters.
+ */
+sealed class ArgumentType(val comparable: Value?) {
+  class SingleValue(comparable: Value?) : ArgumentType(comparable)
+  class SetOfValues(comparable: Value?) : ArgumentType(comparable)
+
+  override fun toString() = "${this.javaClass.name}: $comparable"
+
+  companion object {
+    fun boolean(expression: SqliteParser.ExprContext) =
+        SingleValue(Value(expression, SqliteType.INTEGER, false))
+  }
+}
+
+/**
+ * Represents a full argument as it would appear in the generated code.
+ */
+data class Argument(
+    /** The type to be used in the generated signature. */
+    val argumentType: ArgumentType,
+    /** The text ranges in the original statement this argument replaces */
+    val ranges: List<IntRange>,
+    /** The index into the signature for this parameter */
+    val paramIndex: Int? = null,
+    /** The explicit index the bind arg gave (ie ?3) */
+    val index: Int? = null,
+    /** The explicit name the bind arg gave (:example, @example or $example) */
+    val name: String? = null
+)
+
+/**
+ * All arguments are given a number (even named parameters). Numbering is handed out linearly
+ * over the statement. When a question mark or named argument is observed, it is the number
+ * after the current highest observed number. If a numbered arg shows up then no computation is
+ * needed (other than increasing the highest number seen so far).
+ *
+ * Additionally, android is only capable of taking a string array for its args. And if you
+ * only specify a single argument in your statement but number it ?100 then the string
+ * array will require 99 ignored arguments. Because android only accepts non null strings
+ * we will have to do manual string concatenation for other argument types.
+ *
+ * The argument list stored will be prepared in a way that matches the SQLite rules: Every
+ * arg will be numbered and deduped. The receiver list should be already ordered so all that
+ * remains is deduping and handing out numbers.
+ */
+fun List<Argument>.toSqliteArguments(): List<Argument> {
+  val numberedParameters = ArrayList<Argument>()
+  val numbersToSkip = LinkedHashSet<Int>()
+  var highestNumber = 0
+
+  for (i in 0..size - 1) {
+    if (numbersToSkip.contains(i)) continue
+
+    var argument = get(i)
+    if (argument.index == null) {
+      argument = argument.copy(index = highestNumber + 1)
+    }
+
+    for (j in (i + 1..size - 1) - numbersToSkip) {
```

Comment:
I wonder if there's a better way to do this with iterable combination functions, but good enough for now.